File: ios/Sources/FacesKit/FaceMatcher.swift
import Foundation

final class FaceMatcher {
    struct Candidate {
        let worker: Worker
        let score: Float
    }

    func bestMatch(embedding: [Float], workers: [Worker], threshold: Float) -> Candidate? {
        var best: Candidate?
        for worker in workers {
            let score = cosineSimilarity(embedding, worker.averageEmbedding)
            if score > threshold, score > (best?.score ?? -1) {
                best = Candidate(worker: worker, score: score)
            }
        }
        return best
    }

    private func cosineSimilarity(_ a: [Float], _ b: [Float]) -> Float {
        guard a.count == b.count, !a.isEmpty else { return 0 }
        return zip(a, b).reduce(0) { $0 + $1.0 * $1.1 }
    }
}

File: android/faceskit/src/main/java/ai/unlikeother/faceskit/FaceMatcher.kt
package ai.unlikeother.faceskit

internal class FaceMatcher {
    data class Candidate(val worker: Worker, val score: Float)

    fun bestMatch(embedding: FloatArray, workers: List<Worker>, threshold: Float): Candidate? {
        var best: Candidate? = null
        for (worker in workers) {
            val score = dot(embedding, worker.averageEmbedding)
            if (score > threshold && score > (best?.score ?: -1f)) {
                best = Candidate(worker, score)
            }
        }
        return best
    }

    private fun dot(a: FloatArray, b: FloatArray): Float {
        var sum = 0f
        for (i in a.indices) sum += a[i] * b[i]
        return sum
    }
}
